File: leetcode-submissions/0724-find-pivot-index/solution.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        """
        Time: O(n)
        Space: O(n)
        Intuition: Iterate while hashing each index with its right and left sums. Find an index whose sums are equal.
        """

        left_hash, right_hash = {}, {}
        length = len(nums)
        left_sum, right_sum = 0, 0

        i = 0
        j = length - 1
        while i < length and j >= 0:

            left_hash[i] = left_sum 
            left_sum += nums[i]

            right_hash[j] = right_sum
            right_sum += nums[j]

            i += 1
            j -= 1

        for k in range(length):
            if left_hash[k] == right_hash[k]:
                return k 

        return -1



        

File: leetcode-submissions/0724-find-pivot-index/solution.java
class Solution {
    public int pivotIndex(int[] nums) {
        /**
            prefixsum... for each point do check the sum on both ends... 
         */

        int[] prefixSum = new int[nums.length];
        prefixSum[0] = nums[0];

        for(int i = 1; i < nums.length; i++) {
            prefixSum[i] = prefixSum[i - 1] + nums[i];
        }

        
        for(int j = 0; j < nums.length; j++) {
            int leftSum = j == 0 ? 0 : prefixSum[j - 1];
            int rightSum = prefixSum[nums.length - 1] - prefixSum[j];
            if (leftSum == rightSum) { return j; }
        }
        return -1;   
    }
}
